File: src/com/blackoutburst/game/EntityManager.java
package com.blackoutburst.game;

import com.blackoutburst.bogel.maths.Vector3f;

import java.util.ArrayList;
import java.util.List;

public class EntityManager implements EntityNetworkController {
	
	public final static List<Entity> ENTITIES = new ArrayList<>();
	public static int myId;
	
	public static void init(int id) {
		myId = id;

		ENTITIES.add(new EntityPlayer(id, EntityTypes.PLAYER, new Vector3f(0, 2.80f, 0), new Vector3f(1), new Vector3f(0)));
	}
	
	public static void update() {
		for (Entity e : ENTITIES) {
			e.update();
		}
	}
	
	public static void render() {
		for (Entity e : ENTITIES) {
			e.render();
		}
	}
}

File: client/src/main/kotlin/dev/blackoutburst/game/core/entity/EntityManager.kt
package dev.blackoutburst.game.core.entity

import dev.blackoutburst.game.Main
import dev.blackoutburst.game.maths.Vector2f
import dev.blackoutburst.game.maths.Vector3f
import java.util.*
import kotlin.collections.LinkedHashSet

class EntityManager {
    private val entities: MutableSet<Entity> = Collections.synchronizedSet(LinkedHashSet())

    fun addEntity(entity: Entity) {
        if (entity.id == Main.connection.id) return

        entities.add(entity)
    }

    fun update() {
        entities.forEach {
            if (it.id != Main.connection.id) {
                it.position = it.position.lerp(it.pasSmoothPosition, 0.1f)
                it.rotation = it.rotation.lerp(it.pasSmoothRotation, 0.1f)
            }
            it.update()
        }
    }

    fun render() {
        entities.forEach { it.render() }
    }

    fun setPosition(id: Int, position: Vector3f) {
        if (id == Main.connection.id) return

        entities.find { it.id == id }?.let {
            it.pasSmoothPosition = position
        }
    }

    fun setRotation(id: Int, rotation: Vector2f) {
        if (id == Main.connection.id) return

        entities.find { it.id == id }?.let {
            it.pasSmoothRotation = rotation
        }
    }

    fun removeEntity(id: Int) {
        entities.find { it.id == id }?.let {
            entities.remove(it)
        }
    }
}